aaelf64-morello/aaelf64-morello.rst

```diff
@@ -0,0 +1,831 @@
+
+
+Static linking with Morello
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+A capability has more associated information than a conventional pointer. It has
+extra information, for example base, offset, size and permissions.
+
+Capabilities cannot be statically initialised. Global capability initialization
+when static linking is performed by the runtime at program startup. The
+communication between the static linker and runtime is implementation defined.
+This document describes an implementation implemented based on a table of
+capability descriptions created at static link time, where each capability
+generating relocation results in one entry in the table. When static linking,
+all capability descriptions will be explicitly grouped into a single table of
+capability descriptions where each table entry is a ``struct capdesc`` listed
+below.
+
+In the current LLVM based Morello toolchain, the runtime iterates through each
+``capdesc`` entry creating a capability in the location pointed to by
+cap_location, with the specified base, offset, size and permissions given by the
+entry. To aid in the finding of the capability descriptions table the linker
+emits two symbols to denote the start and end of the table:
+``__cap_relocs_start`` and ``__cap_relocs_end`` respectively. The capability
+descriptions table is placed inside the ``__cap_relocs`` section.
+
+.. code-block:: c
+
+    struct capdesc
+    {
+        void*__capability cap_location;
+        void* base;
+        uint64_t offset;
+        uint64_t size;
+        uint64_t permissions;
+    };
+
+The permission bits of a capability constructed for a ``capdesc`` entry is the
+inverse of the ``permissions[17:0]`` field in the ``capdesc`` entry.
+Additionally, the MSB (bit 64) of the ``capdesc`` ``permissions`` field is set
+for Executable symbols to indicate that the PCC is to be used to construct the
+Capability.
+
++-------------------+---------------------------------+
+| Permission        | Encoding                        |
++-------------------+---------------------------------+
+| Executable        | ``0x8000000000013DBCULL``       |
++-------------------+---------------------------------+
+| Read-Write Data   | ``0x8FBEULL``                   |
++-------------------+---------------------------------+
+| Read-Only Data    | ``0x1BFBEULL``                  |
++-------------------+---------------------------------+
+
+When a Morello-capable assembler sees a ``.capinit`` instruction, it reserves a
+16-byte (128 bits) location (``fragment``) and  generates a
+``R_MORELLO_CAPINIT`` relocation for the linker to create a capability in the
+``fragment``. The assembler may use the ``fragment`` with the following format
+to give out size hints for the linker to use before processing the relocation.
+
++----------------+--------------+
+|  64-bit: empty | 64-bit: size |
++----------------+--------------+
+
+This size hint if not superseded by more accurate information will be
```

Comment:
This size hint will be incorporated.. field, if not superseded by more accurate information.